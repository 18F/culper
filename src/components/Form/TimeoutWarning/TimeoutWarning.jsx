import React from 'react'
import { connect } from 'react-redux'
import { i18n } from '../../../config'
import { api } from '../../../services'
import { updateApplication } from '../../../actions/ApplicationActions'
import { logout, tokenError } from '../../../actions/AuthActions'
import { saveSection } from '../../../middleware/history'
import AuthenticatedView from '../../../views/AuthenticatedView'
import Branch from '../Branch'
import Modal from '../Modal'
import Show from '../Show'
import Svg from '../Svg'

const roundUp = (num, precision) => {
  return Math.ceil(num * precision) / precision
}

const minutes = (ms) => {
  return parseInt(roundUp(ms / (60*1000), 1), 10)
}

const seconds = (ms) => {
  return parseInt(roundUp(ms / 1000, 1), 10)
}

export class TimeoutWarning extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      showWarning: this.props.showWarning,
      interval: props.interval || 60 * 1000,
      countdown: 60
    }
    this.onDismiss = this.onDismiss.bind(this)
    this.onClick = this.onClick.bind(this)
    this.tick = this.tick.bind(this)
    this.timer = null
  }

  componentDidMount () {
    this.resetInterval()
  }

  componentWillUnmount () {
    window.clearInterval(this.timer)
  }

  tick () {
    const now = new Date().getTime()
    const last = this.props.lastRefresh
    const timeoutInMilliseconds = this.props.timeout * 60 * 1000
    const diff = now - last
    if (seconds(diff) >= seconds(timeoutInMilliseconds)) {
      // Session has expired
      this.props.dispatch(tokenError())
    } else if (this.state.showWarning) {
      // Ticks are to count down by seconds now
      this.setState({ countdown: seconds(timeoutInMilliseconds - diff) })
    } else {
      // Ticks should be checking if in the last minute of the session
      if (minutes(diff) >= this.props.timeout - 1) {
        const application = this.props.app
        const section = this.props.section.section
        const subsection = this.props.section.subsection
        const dispatcher = this.props.dispatch
        this.setState({ showWarning: true, interval: 1000, countdown: seconds(timeoutInMilliseconds - diff) }, () => {
          // Save the form
          saveSection(application, section, subsection, dispatcher)
          // Change the timer interval
          this.resetInterval()
        })
      }
    }
  }

  onDismiss () {
    this.refreshToken()
  }

  onClick () {
    this.setState({ showWarning: false, interval: 60 * 1000, countdown: 60 }, () => {
      // Change the timer interval
      this.resetInterval()
      // Refresh the token
      this.refreshToken()
    })
  }

  resetInterval () {
    if (this.timer) {
      window.clearInterval(this.timer)
    }
    this.timer = window.setInterval(this.tick, this.state.interval)
  }

  refreshToken () {
    const self = this
    api.refresh().then(r => {
      api.setToken(r.data)
      if (r.data === '') {
        self.props.dispatch(tokenError())
      } else {
        self.props.dispatch(updateApplication('Settings', 'lastRefresh', new Date().getTime()))
      }
    }).catch(() => {
      api.setToken('')
      self.props.dispatch(tokenError())
    })
  }

  message () {
    return i18n.t('application.timeout.message').replace('{time}', this.state.countdown)
  }

  render () {
    return (
      <div className="timeout-warning">
        <Modal show={this.state.showWarning}
               onDismiss={this.onDismiss}>
          <Svg src="/img/timeout-icon.svg" />
          <h2>{i18n.t('application.timeout.title')}</h2>
          <p>{this.message()}</p>
          <button title={i18n.t('application.timeout.button')} aria-label={i18n.t('application.timeout.button')} onClick={this.onClick}>
            <div>
              <span>{i18n.t('application.timeout.button')}</span>
              <i className="fa fa-arrow-circle-right" aria-hidden="true"></i>
            </div>
          </button>
        </Modal>
      </div>
    )
  }
}

TimeoutWarning.defaultProps = {
  lastRefresh: null,
  showWarning: false,
  timeout: 15,
  dispatch: (action) => {}
}

function mapStateToProps (state) {
  const section = state.section || {}
  const app = state.application || {}
  const settings = app.Settings || {}
  return {
    section: section,
    app: app,
    lastRefresh: settings.lastRefresh
  }
}

export default connect(mapStateToProps)(AuthenticatedView(TimeoutWarning))
