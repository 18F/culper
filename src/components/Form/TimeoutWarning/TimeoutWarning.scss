.timeout-warning {
  .modal {
    background-color: rgba(255, 255, 255, 0.97);

    .modal-content {
      max-width: 64rem;
      text-align: center;
      background: transparent;
      border: none;
      box-shadow: none;

      h2 {
        margin-top: 1.3rem;
      }

      button {
        padding: 1.7rem 3rem;

        i {
          margin-left: 1rem;
        }
      }
    }
  }
}
