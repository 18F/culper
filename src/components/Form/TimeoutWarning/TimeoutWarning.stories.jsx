import React from 'react'
import { storiesOf } from '@storybook/react'

import { TimeoutWarning } from './TimeoutWarning'

storiesOf('TimeoutWarning', module)
  .add('default', () => (
    <TimeoutWarning
      showWarning
    />
  ))
