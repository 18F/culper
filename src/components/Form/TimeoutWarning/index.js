import TimeoutWarning from './TimeoutWarning'
export default TimeoutWarning
