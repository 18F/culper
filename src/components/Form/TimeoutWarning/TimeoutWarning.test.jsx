import React from 'react'
import MockAdapter from 'axios-mock-adapter'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import TimeoutWarning from './TimeoutWarning'

describe('The timeout warning component', () => {
  // Setup
  window.token = 'fake-token'
  const middlewares = [ thunk ]
  const mockStore = configureMockStore(middlewares)

  it('not displayed if outside threshold', () => {
    const store = mockStore({ application: { Settings: { lastRefresh: new Date() } }, authentication: { authenticated: true, twofactor: true } })
    const component = mount(<Provider store={store}><MemoryRouter><TimeoutWarning timeout="15" /></MemoryRouter></Provider>)
    console.log(component.html())
    expect(component.find('.timeout-warning').length).toEqual(1)
    expect(component.find('.modal').length).toEqual(0)
  })

  it('displayed when inside threshold', () => {
    const now = new Date()
    let lastRefresh = new Date(now.setMinutes(now.getMinutes() - 14))
    const store = mockStore({ application: { Settings: { lastRefresh: lastRefresh } }, authentication: { authenticated: true, twofactor: true } })
    const component = mount(<Provider store={store}><MemoryRouter><TimeoutWarning timeout="15" showWarning={true} /></MemoryRouter></Provider>)
    console.log(component.html())
    expect(component.find('.timeout-warning').length).toEqual(1)
    expect(component.find('.modal').length).toEqual(1)
  })
})
