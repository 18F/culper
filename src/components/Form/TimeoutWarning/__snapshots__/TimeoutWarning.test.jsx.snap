// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The timeout warning component matches the snapshot 1`] = `
<div
  className="timeout-warning"
>
  <div
    aria-hidden="false"
    className="modal"
    onClick={[Function]}
    role="dialog"
  >
    <div
      className="modal-wrap"
    >
      <div
        className="modal-content"
        onClick={[Function]}
      >
        <img
          alt="Scalable vector graphic"
          className="svg"
          src="/img/timeout-icon.svg"
        />
        <h2>
          Are you still working?
        </h2>
        <p>
          Let us know below or you will be logged out in 60 seconds.
        </p>
        <button
          aria-label="Yes I'm still working"
          onClick={[Function]}
          title="Yes I'm still working"
        >
          <div>
            <span>
              Yes I'm still working
            </span>
            <i
              aria-hidden="true"
              className="fa fa-arrow-circle-right"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
